### ISSUE TITLE
How to use custom metrics?

### ISSUE THREAD

**Issue by thisroot** [zsais/go-gin-prometheus#25]
Hi, i read sourse code and could not understand how to use custom metrics? Does exist way do it, or i should rewrite this library?

**Comment by nizarayari** [comment]
Any update on this issue? It seems you can add custom metrics but the question is how to trigger them?

Thanks in advance!

**Comment by thisroot** [comment]
@nizarayari Yes, you should use gin metrics with you custom mectrics build with standart prometheus library methods.  Gin prometheus library collect base metrics and bind it into common metrics report

**Comment by lmcdasm** [comment]
ok.. so im a bit stumped here.

i have the custom metrics implemented (so i can see them show up when i have my server running):

![image](https://user-images.githubusercontent.com/9358276/131207505-3cf3e4db-e114-4ad8-b6b3-8112372374b3.png)

However, even with looking with the answer provided to nizarayari, i dont see how to apply the .Inc() method to the custom metric.

Here is my code snippet:

`func main() {
      r1 := gin.Default() // Main Server
      r2 := gin.Default() // A server where Prometheus itself runs on 
      // Optional Custom Metrics List
        customMetrics := []*ginprometheus.Metric{
                &ginprometheus.Metric{
                        ID:          "9001",          // required iD
                        Name:        "test_metric",   // required string
                        Description: "A test metric", // required string
                        Type:        "counter",
                },
                &ginprometheus.Metric{
                        ID:          "9002",          // required ID
                        Name:        "test_metric_2", //required string
                        Description: "test 2 metric",
                        Type:        "summary",
                },
                // TYPES: counter, counter_vec, gauge, gauge_vec, histogram, histogram_vec,summary, sum_vec
        }
        p := ginprometheus.NetPrometheus("gin", customMetrics)
        r1.Use(p.HandlerFunc())
        p.SetMetricsPath(r2)
        r1.GET("/alive", func (c *gin.Context) {
                      // I want to increase the counter on this hit
                      // This is 'wrong" , but iterating i know its the first it the sllice
                      test_cnt := p.MetricsList[0]
                      test_cnt.Inc()`
                      c.JSON(http.StatusOK, gin.H{ 
                                  "code": http.StatusOK, 
                                  "status": "alive", 
         })
         
         go func () { r2.Run(":8443") } ()
         err := r1.Run(":5000")
         if err != nil { panic (err) }
}

I have tried as well looking at the other example, however the prometheus.Register(numErrors) example never ends up showing up in the gin metrics list (as the test ones outlined above are).

any pointers would be wonderful!
D

**Comment by snowlyg** [comment]
I change `MetricsList []*Metric` to `MetricsList map[string]*Metric`, when i use the custom metrics. Make a easy way to get  custom metric.

```go
customMetrics := []*ginprometheus.Metric{
	&ginprometheus.Metric{
		ID:	"1234",				// optional string
		Name:	"test_metric",			// required string
		Description:	"Counter test metric",	// required string
		Type:	"counter",			// required string
	},
	&ginprometheus.Metric{
		ID:	"1235",				// Identifier
		Name:	"test_metric_2",		// Metric Name
		Description:	"Summary test metric",	// Help Description
		Type:	"summary", // type associated with prometheus collector
	},
	// Type Options:
	//	counter, counter_vec, gauge, gauge_vec,
	//	histogram, histogram_vec, summary, summary_vec
}
p := ginprometheus.NewPrometheus("gin", customMetrics)
// add data to custom metrics
m := p.MetricsList["1234"].MetricCollector.(prometheus.Counter)
m.Inc()
```

**Comment by lmcdasm** [comment]
Thanks @snowlyg
